### ISSUE TITLE
[Update] Iridium is now a Firefox only extension

### ISSUE THREAD

**Issue by ParticleCore** [ParticleCore/Iridium#737]
Some of you might not have noticed (insert the biggest sarcasm face here) but I haven't been able to maintain this extension for a while and this is a choice I made as a result of no longer having enough free time as I did once before.

However, I never gave up on this extension and have recently been slowly (OK, maybe EXTREMELY slowly) experimenting better ways to keep this easier for me to keep developing and maintaining and I've reached a decision that will leave a lot of users upset - me included - but it's a change that has to be done as a result of my very limited to non-existent free time.

The extension will become a Firefox webextension only in the next version, this means no more userscripts and no more Chrome extension. Not because I want to exclude those platforms, but because they lack a very important extension API that only Firefox has at the moment, that api being [filterResponseData](https://developer.mozilla.org/docs/Mozilla/Add-ons/WebExtensions/API/webRequest/filterResponseData).

What this API allows is the modification of any resource before it is loaded in the user's browser. This has a huge advantage over what I've been doing so far because unlike my past methods I no longer have to rely on finding "holes" in the YouTube source code in order to manipulate the page navigation, player behavior, etc.
With this API I can easily modify the source code directly in any way I desire which cuts down a lot of headache and time spent with the previous method (there were times where it would take me one month just to make one feature work again).

Unfortunately this API only exists in Firefox. Chrome never adopted it - I believe they tried but it's very buggy and has been left forgotten since their first attempt - and userscript managers would need to expand or implement this permission in their framework. I'm going to request them to implement this, I know Tampermonkey is already trialing webRequest, which the API I need is a subset of, so it might be possible in the future to extend support to userscript as well, but no promises since it depends entirely on the developers' decision.

I've been focusing on the bad news mostly so far to get those out of the way, so now is time for the good news.

- As I explained, the biggest one - for me of course - will be the significant reduction in developing and maintaining code that is subject to YouTube changes in their source code

- The other big one is Iridium will be able to implement the video download feature since it will no longer be distributed as a Chrome extension. However, this will be a simple feature, there will be no video packaging/conversion, no encoding whatsoever, the feature will let the user save locally the stream files YouTube provides, nothing more. These include video (without audio), audio and video thumbnail when available

- Iridium will be lighter since it will be taking advantage of more advanced features that are native to the browser and will be smaller in size (hey, every kb counts in today's day and age)

I plan to jump the major version to 1 and restart the rest, so this change will be released on version 1.0.0 and I'm going to leave the major version 0 just for any-scenario sake such as backporting something to previous versions, if it ever happens.

I thought I had more to write, but I've been staring at this draft for more than 10 minutes and I can't recall anything else. I'll probably remember after posting as usual, but if that's the case I'll update this with the relevant information.

As always I am very sorry that it has been taking this long, but as I always explained I couldn't put this over my professional life unless this was a reliable source of income. I thank everyone that has been sticking around - even though the extension barely works - and more importantly to those that have been trying to help other users in this repository.

**Comment by ParticleCore** [comment]
> 
> 
> hey i've long uninstalled this buggy thing
> 
> please feel free to just cancel it. i'll be very happy. i once wrote a long list of the duplicate & outdated things that this extension does, and nothing was ever fixed
> 
> that was long before this point: #570
> 
> was doing some github issues cleanout, and i guess perfect timing for https://github.com/ParticleCore/Iridium

I apologize if you feel that way. I checked what you were referring to and all I could see was the following in your only opened issue #570:

> video does not play automatically when that option is turned on in settings, how troubleshoot?

You've recently removed that from your issue and I still cannot find the list you mentioned.

The development of the extension (including bug fixes) has been on hold for quite some time, it wasn't for the lack of wanting to fix things, but there are surely working alternatives that would suit you best, you would do right to uninstall this one if it is not meeting your expectations.

**Comment by anowlcalledjosh** [comment]
> The extension will become a Firefox webextension only in the next version, this means no more userscripts and no more Chrome extension.

That's fair enough. Since some features of the extension still work fine, will the current version continue to be available on the Chrome Webstore?

**Comment by ParticleCore** [comment]
> 
> 
> > The extension will become a Firefox webextension only in the next version, this means no more userscripts and no more Chrome extension.
> 
> That's fair enough. Since some features of the extension still work fine, will the current version continue to be available on the Chrome Webstore?

The current versions that are available will not be removed, they will remain as they are, the only change will be done on the Firefox WebExtension that will upgrade to version 1.0.0 with these new changes when they become available.

**Comment by wrested** [comment]
Hey thank you very much for the work you put in on Iridium while it was around.  It was easily the best YouTube tweak I've ever used and I'm still using the features that still work today.  It will be missed!  (created an account here just to let you know btw) 

Good luck with the Firefox fork.  I've been meaning to give that browser a try again to see how it's changed in the 10 years since I've had it installed and I'll make sure to give Iridium a shot when I do.

**Comment by dominateeye** [comment]
Do you know if Vivaldi or any other Chromium-based independent browsers support filterResponseData or webRequest? If so, could new Iridium still work on them? I soured on Firefox after they implemented that DRM thing that the WWW Consortium was pushing a few years back, and Vivaldi has a lot of features I really like, so I'd prefer not to have to switch browsers if I don't have to.

**Comment by ParticleCore** [comment]
> 
> 
> So it's just a mirror?

Correct, one is hosted on AMO and the other is hosted here in this repository.



> 
> 
> Do you know if Vivaldi or any other Chromium-based independent browsers support filterResponseData or webRequest? If so, could new Iridium still work on them? I soured on Firefox after they implemented that DRM thing that the WWW Consortium was pushing a few years back, and Vivaldi has a lot of features I really like, so I'd prefer not to have to switch browsers if I don't have to.

The link I posted shows that only Firefox supports this API at the moment. Unless Chrome adopts the same API no other Chromium based browser will support that API.

**Comment by wadawada** [comment]
Thanks for the update
I had been struggling using the script as it lacks maintenance and started breaking apart for chrome . I finally found an alternative few days ago. 
Although I quite like tags function on Firefox, I still can't convenience myself to completely switch to Firefox. However, I look forward to seeing how new Iridium will work on Firefox (maybe with new features?)

**Comment by thepcwiz101** [comment]
Posted report to bugs.chromium.org: 
![image](https://user-images.githubusercontent.com/31826503/64919004-47590680-d76b-11e9-8052-523c3c8fd298.png)

**Comment by ganego** [comment]
Well kinda sucks because I use Waterfox (which does not have the API), but I see your reasoning. Good luck and thank you for updating it.

**Comment by thepcwiz101** [comment]
@ganego it turns out that there was a report already made for adding support for this but they have no plans on adding it. This is very frustrating tbh. Google has all this time to implement more bloat and features such as sync but they can't add one simple thing like this? I swear though recent decisions google has made is really pushing me closer and closer to installing firefox.

**Comment by DioBrando141** [comment]
great to hear this. Came back to FF since Quantum, and loving my time with it, more so when webrender launched. it was also necessary for me to migrate to another extension, but i miss various functions only Iridium had, so i'm excited for its return.

**Comment by Ferm93** [comment]
Great to hear you found a way to keep updating the extension and did not have to just drop it.
Can't wait for future updates!

Thank you for your hard work man

**Comment by ImprovedTube** [comment]
We can collaborate

**Comment by BobbyWibowo** [comment]
Yay, no autoplay is back! 💃🕺

**Comment by ParticleCore** [comment]
> 
> 
> Yay, no autoplay is back! dancerman_dancing

It's already working, but it is a very very buggy version, however anyone can try it out already from branch 1.0.0 https://github.com/ParticleCore/Iridium/tree/1.0.0

**Comment by thepcwiz101** [comment]
Also if the no autoplay option isn't working right for you due to it being buggy there's an extension called AutoplayStopper that allows to block all websites from auto playing video. You can also customize the rules in the extension to exclude websites.

**Comment by ParticleCore** [comment]
Please don't use this issue and repository to publicize other extensions

**Comment by thepcwiz101** [comment]
@ParticleCore Just trying to help people out.

**Comment by Corruptinglyneedful** [comment]
Did you link to it?

**Comment by thepcwiz101** [comment]
@Corruptinglyneedful Nope i purposely just gave the extension name and the dev forgot one of the names the extension went by on firefox but i deleted the comment regardless. I posted that comment to help people just in case @ParticleCore doesn't get any time to fix the issues.

**Comment by Corruptinglyneedful** [comment]
Mmm, I don't see an issue as long as you don't link to it but that's just my two cents.

@ParticleCore Has there been any updates to the extension since I downloaded it in our last conversation?

**Comment by BobbyWibowo** [comment]
> It's already working, but it is a very very buggy version, however anyone can try it out already from branch 1.0.0 https://github.com/ParticleCore/Iridium/tree/1.0.0

Aye aye captain! I was actually already using it before I sent the comment, lul
I only noticed sometimes HQ thumb didn't work on my end

**Comment by thepcwiz101** [comment]
@Corruptinglyneedful  i probably should have just said in my comment: "If you have any problems and the dev can't fix it for a little while then lookup an extension that blocks autoplay on websites." It can be used alongside iridium so it really isn't bringing people away from using his extension lol.

**Comment by BobbyWibowo** [comment]
@Corruptinglyneedful Just look out at the 1.0.0 branch. It's on dev so I don't think he will bother sign the extension everytime he pushes a commit, so you will have to clone the branch and build/sign it yourself with `web-ext` (at least that's what I did)

There aren't really that much in it for now tho, but no autoplay is in it. And it's a proper no autoplay, as in it actually prevents the video element from loading & buffering in the background.

**Comment by ParticleCore** [comment]
@BobbyWibowo Autoplay is in it already, open the Iridium settings and turn it off.

![ss1](https://user-images.githubusercontent.com/9222661/65360605-1146d700-dbf0-11e9-9f63-ca2c9f9bf967.png)

**Comment by BobbyWibowo** [comment]
@ParticleCore I said no autoplay, cause that's what I want it to do, not autoplaying 😂

**Comment by ParticleCore** [comment]
> 
> 
> @ParticleCore I said no autoplay, cause that's what I want it to do, not autoplaying joy
> It's basically just my way of phrasing when "Play videos automatically" is unticked
> And I mentioned it being proper cause other solutions tend to just pause the video while still letting it to buffer

I get it now, it was so confusing at first. Yeah, it's working quite nicely again FINALLY. I'm gonna focus on organizing the base code structure so I can move on to adding more features and also add a autoplay button on the page for easier access.

**Comment by ParticleCore** [comment]
> 
> 
> > It's already working, but it is a very very buggy version, however anyone can try it out already from branch 1.0.0 https://github.com/ParticleCore/Iridium/tree/1.0.0
> 
> Aye aye captain! I was actually already using it before I sent the comment, lul
> I only noticed sometimes HQ thumb didn't work on my end

Also the HQ thumbnail should be working all the times, except when there isn't one available, then it shows the grey 3 dots default thumbnail. Not much I can do about that right now.

**Comment by BobbyWibowo** [comment]
> Also the HQ thumbnail should be working all the times, except when there isn't one available, then it shows the grey 3 dots default thumbnail. Not much I can do about that right now.

Yeah, it was showing the three dots thingy on some videos yesterday. Probably one of those non-HD videos? Not sure, I wasn't really paying attention. Though I always believed `maxresdefault` thumb would always be filled with something, regardless of whether the videos were HD or not (cause like, by the name, I'd assumed it'd just fill it with the highest res of thumbs available).

Also would love to see the option to enforce video quality again 👍

**Comment by Corruptinglyneedful** [comment]
Yah idk how to do that.  I'm a simple user.

**Comment by BobbyWibowo** [comment]
Yeah, that's fair. Just gotta be a bit more patient I suppose then? I mean heck, we've waited for almost a year, what's a few more days/weeks, lul

Anyway, if you feel comfortable for a bit of manual tinkering, you can also simply load the `/src/Webextension` directory as a temporary add-on by following the instructions here: https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Temporary_Installation_in_Firefox.
Don't forget to download/clone the `1.0.0` branch somewhere first.

But this method will only make the extension stay during that session. As in, once you close Firefox, you'll have to re-do the steps again.

---

You will only need to sign it with `web-ext`, if you want it to install normally on a regular Firefox edition, and will stay even after you've restarted the browser.
It's documented here, if you're still interested: https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Getting_started_with_web-ext.

It may end up being more work than worth if you haven't used the tool before, or if you don't really plan on delving any further into making Firefox's extensions, so I don't really recommend it.

I personally don't think it's much of an extra work for Particle to upload a signed test build every once in a while, but it still IS some extra work, so I won't comment further.

**Comment by Corruptinglyneedful** [comment]
Hey thanks for that information.  Nope, got no interest in developing my own extensions, but the information could come in handy for the future or for others.

**Comment by ng4ever** [comment]
Thanks for keeping with this add-on! Greatly appreciated! Plan on donating! 

So far so good!

**Comment by thepcwiz101** [comment]
Since the userscript will no longer be developed the "userscript" tag should be removed.

**Comment by ParticleCore** [comment]
@thepcwiz101 Correct, but not before the refactored version is out. I'll do a total clean-up by then

**Comment by thepcwiz101** [comment]
@ParticleCore oh ok lol

**Comment by Onurtag** [comment]
Glad to see you are back.  
With the api you mentioned we might finally be able to get #705 done without using bandaid fixes.

**Comment by ParticleCore** [comment]
@Onurtag Totally possible since I can change the source code to remove the loudness equalization algorithm use completely

**Comment by erkinalp** [comment]
> Chrome never adopted it - I believe they tried but it's very buggy and has been left forgotten since their first attempt -

They never tried. This API is not included in Chrome as it faciliates adblocking.

**Comment by TheDoctor40** [comment]
Dang. I'm glad that at least the already existing version will remain working on Chrome. One thing I don't like as much is that suddenly my Firefox version is severely stripped of it's functions. I'd much rather have the old version untill the new one is ready instead of having the new version right now. Luckily you can install older versions using the Firefox Add-ons store so I'll just do that and turn off auto-update for Iridium untill it has more features which I use.

**Comment by BobbyWibowo** [comment]
@TheDoctor40 If you have a userscript manager (e.i. Tampermonkey or Greasemonkey), you can still use the userscript version alongside the new Firefox extension.
```
https://github.com/ParticleCore/Iridium/raw/37d27616cbfb70b0296b9a16d41ff31986d58575/src/Userscript/Iridium.user.js
```
That's the latest userscript from commit https://github.com/ParticleCore/Iridium/commit/37d27616cbfb70b0296b9a16d41ff31986d58575 (0.2.5).

I'm on Firefox and have always been using both at the same time.

The userscript's "Video stays always visible while scrolling" and "Default video quality" are still working, and I can't really go without them yet.
While the new extension will properly handle disabling autoplay & max res thumbnails.

**Comment by Acid-Crash** [comment]
Hi @ParticleCore.
Wanted to give you thumbs up for the announcement.
Waiting for all future updates!
Also I'm really sad for all the negative a hate from some users that don't bother to read the announcement before firing a new ticket-bugreport with "WTF, it is broken after the update".
Guess some people will never learn...

**Comment by JordyValentine** [comment]
I completely understand why you are updating and taking advantage of the API rather than hooking in as you've done previously. My only suggestion is that maybe this latest version could have been released as a beta, I feel like that would have avoided a lot of the backlash.

Personally the old version has worked fine for me, so I'll stick with that until the new one is a bit more feature complete, I just wish I'd followed the project more closely so I would have seen the notice earlier.

**Comment by Corruptinglyneedful** [comment]
There was going to be backlash regardless.

**Comment by erkinalp** [comment]
@ParticleCore Can you add a "download video" button on non-premium (always present on CC licensed videos and opt-in with standard ones)? Iridium conflicts with other extensions adding such buttons).

**Comment by ParticleCore** [comment]
@erkinalp I don't understand what you are asking, so please open a new issue and detail better what you are trying to explain.

**Comment by JordyValentine** [comment]
@CorruptinglyneedfulI
I agree there still would have been some backlash, but making the update optional would have meant it was say, 10% of the user base upset, rather than having 100% percent of install break and wipe their settings.

**Comment by LaquedemPF** [comment]
Glad to see that Iridium could have a future after all that issues that were before with the popout player & people locking the addon out of the chrome store back then...
Yet it caught me offguard that the addon get's recoded and that much functionality is gone. I thought at the first moment that the addon broke due some YouTube Update, and so do probably a lot of people that just want to use the addon & don't step in the material and look it up on Github. So i wonder if you might want to notify users at some way about the recode before you lose both hundreds of users & the addon´s "Featured" State in AMO?
https://addons.mozilla.org/de/firefox/addon/particle-iridium/reviews/
https://addons.mozilla.org/de/firefox/addon/particle-iridium/statistics/?last=30

**Comment by thepcwiz101** [comment]
@ParticleCore I recommend reverting the firefox extension to the previous version. The rebuilt code must be matured and considered stable before a proper release no matter how long it takes. Ik an extension (not to be mentioned here) that has 2 store versions one is the beta and the other is the stable build. I hope this helps you develop your extension better.

**Comment by jrial** [comment]
Just adding my $0.02 that this release could've been handled better. Apart from the autoplay being broken, there was nothing wrong with the previous version, which had tons of functionality.

It's great that autoplay works again, but frankly, Firefox kind of fixed that for my use case a long time ago by not playing videos opened in new, inactive tabs. On the other hand, I now lost the mini player that remains on screen while scrolling, which was the main selling point for me. Furthermore, reading the reviews, I see people complaining about the settings being wiped: downgrading to previous version doesn't bring back the blocklist people spent so much time building, and after a downgrade, the settings get reset every time the browser restarts.

This release came out of nowhere, replaced a mostly working piece of software people relied on with something utterly devoid of functionality, caused data loss, and didn't warn its users about any of this. All of this could've been avoided by simply creating a second "beta" or "preview" release and informing the users through the "extension has been updated" news page.

There's a lesson in there. I hope it's taken into account the next time you're faced with a similar choice.

Oh, well, it's done now. I guess I'll just have to wait for the features to gradually reappear. Bummer about the chrome version though, as it'll be only a matter of time until the addon becomes entirely useless on that platform thanks to Google's constant changes to the YT platform.

**Comment by ParticleCore** [comment]
Gonna leave this as the final say on this topic and lock the issue to prevent further repetition.

@jrial 
I had over 400 users reporting that most of the features of the extension were no longer working, that's 400 more than the ones that said that most of the features of the extension were still working.
On my side of things most of the features were indeed not working any more and without any other way or time to ask the remaining users to confirm this I only had the information that most features were no longer working.

I could've made a beta version, I could have updated the previous version with a pending breaking change warning, all of this could've been handled differently. Which it did, twice before, and in the end it still resulted in a lot of users complaining about the changes - because most don't like changes - others complaining that it was taking too long for the release and they'd prefer to have featured releases (release with every new feature implemented) and all of that. I've done it before, with this extension and the one before it.
The result? The exact same result as the one right now: users complaining.

Given the end result is still the same regardless of the release method and given that I have almost no free time to develop the extension it is only logical that I'd chose to go with the method that gives me the least work; fresh start right into stable.

Remember, at this point I still only know that most features don't work.

Then the release is out and guess what, apparently for other users most features were still working. How do I know this? Because of the users complaining about it after the fact.

I've made a few announcements in past issues informing of the situation that lead to this, I've yet to find anyone saying that most of the features were still working.

Granted GitHub issues are not the best way to reach the user base, but I had nothing to suspect that there were users with most features working as such I never had any need to reach out to other users for that feedback.

I do appreciate all the feedback even after the fact, although it is not productive, I find it the only thing it helps is for others to vent out because they were upset.

What about me? Like I said before plenty of times, I do this as a hobby for myself and share publicly for others that'd like to enjoy YouTube better. Donations are welcome, but they don't dictate the direction the extension takes, I've had users emailing me they'd pay to add features and all I reply back is that I add the features that I find to be useful and I appreciate the offering, but aside from donations the feature implementations are free and not pay-based.

I don't have a lot of free time, I do this when I find a chance and what the donations amount to is a nice pocket change at the end of the month, but it is not by any means any income that would allow me to drop my work and focus on this 100%. I'd like for that to be possible, but that's not the reality at the moment.

Users got upset about this? I already apologized. Settings got lost? There's a way to get them, but honestly that's up to the users now. In all honesty I'm just tired of hearing people complaining about this in a way that is not constructive. Especially when users recognize that they are just repeating what others already said, what's the point? That's a rhetorical question by the way.

If you (not "you you", but "general you") don't like the extension - be it at the current or previous state - then it is not doing what it was meant to do: improve user experience on YouTube. If that is the case then find another extension that will help achieve that goal better, this is not a loyalty-based extension, users should not be bound to it.

That's the lesson in here.

I'd like to end this with a more positive note: I appreciate everyone's feedback and interest in the extension, I'm glad to see it helps others as well as it does with me and I plan on adding most, if not all features from the previous version.

Also features the previous version did not have before will be implemented in this version.
